---
title: "TRIAL"
teaching: 10
exercises: 0
questions:
- "Why sequence RNA?"

objectives:
- "Understanding the primary aims of this workshop"
keypoints:
- XXX
---

---
output: 
   html_document:
   code_folding: hide
---

#### Important links
- Analyzing RNA-seq data with DESeq2
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

- RNA-seq workflow: gene-level exploratory analysis and differential expression
https://bioconductor.org/help/course-materials/2017/CSAMA/labs/2-tuesday/lab-03-rnaseq/rnaseqGene_CSAMA2017.html

- Pre-processing of RNA-Seq daatsets (Used by Tracy)
https://compbiocore.github.io/deseq-workshop-1/assets/deseq_workshop_1.html


### Load the R-libraries
```
suppressMessages({
  library("DESeq2")
  library("edgeR")
  library("limma")
  library("RColorBrewer")
  library("gplots")
  library("ggplot2")
  library("EnhancedVolcano")
  library("factoextra")
  library("devtools")
  library("rstudioapi")
  library("dplyr")
  library("tibble")
  library("tidyverse")
  library("pheatmap")
  library("biomaRt")
  library("annotables")
  library("org.Mm.eg.db")
  library("biobroom")
  library("clusterProfiler")
  #library("pathfindR")
})


```

### Set the current working directory
```
current_path <- getActiveDocumentContext()$path 
setwd(dirname(current_path ))
```

### Import the Gene-count matrix (Generated by nfcore-rnaseq)
- As input, the DESeq2 package expects count data, e.g., from RNA-seq or another high-throughput sequencing experiment, in the form of a matrix of integer values. 
- The value in the i-th row and the j-th column of the matrix tells how many reads can be assigned to gene i in sample j. 
- The values in the matrix should be un-normalized counts or estimated counts of sequencing reads (for single-end RNA-seq) or fragments (for paired-end RNA-seq). 
- The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should NOT be used as input.
- Next we read in the count matrix generated by the "nfcore-rnaseq" pipeline.
- We have actually generated a matrix for a sub-set of the data from a specific region of the genome (chromosome 18). This was done with consideration to the time required for mapping and other processing steps involved.
- We will be using this processed - subset matrix as part I of our analysis. 
- We will be re-doing the analysis with a FULL matrix which was downloaded from the original manuscript as part II. We can then use this full data for identifying gene functional categories of interest.

```
# Part I
#counttable_original<-read.delim("subset_matrix.txt", header=T, row.names=1) 

# Part II
counttable_original<-read.delim("GSE81082_count_matrix_ENSIDs_symbols_nr.txt", header=T, row.names=1) 

# View the count matrix
#View(counttable_original)

# Gene symbol as the identifier (when compared to ENSG ID)
counttable<-counttable_original[,c("Symbol","WT1","WT2","WT3","KO1","KO2","KO3")]
row.names(counttable) <- NULL
# Convert Column  'GeneSymbol' to rowname)
rownames(counttable) <- counttable$Symbol
counttable<-counttable[,c("WT1","WT2","WT3","KO1","KO2","KO3")]
#View(counttable)


# There is that question of redundancy? Many genes same ENSG?- To be thought about and addressed

```

### Exploratory analysis and visualization

#### DESeq2 object: DESeqDataSet
- The object class used by the DESeq2 package to store the read counts and the intermediate estimated quantities during statistical analysis is the DESeqDataSet, which will usually be represented in the code here as an object dds.
- Please see the [guide](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) for package information.
- We will then define a condition variable to associate the individual columns (samples) in the matrix to their appropriate experimental condition (either 'Wild-type' or 'Knockout').
- We then prepare the DESeq2 object with `design = ~1`. A design of ~1 is used for no experimental design and is useful for exploring QC of the data (not for DE).

```

# Define a condition variable 
condition=c("Wild","Wild","Wild","KO","KO","KO")
meta <- data.frame(row.names=colnames(counttable),condition)
#View(meta)


dds <- DESeqDataSetFromMatrix(countData = counttable, 
                              colData = meta, 
                              design = ~1)
```

- Blind dispersion estimation is not appropriate if one expects that many or the majority of genes (rows) will have large differences in counts which are explainable by the experimental design, and one wishes to transform the data for downstream analysis. 
- In this case, using blind dispersion estimation will lead to large estimates of dispersion, as it attributes differences due to experimental design as unwanted noise, and will result in overly shrinking the transformed values towards each other.
- This [tutorial](https://compbiocore.github.io/deseq-workshop-1/assets/deseq_workshop_1.html) provides a nice deeper explanation of these concepts.



#### Box plot

```
boxplot(log2((counttable)+1),las=3, col="red")
```

#### Data transformations
- For visualization or clustering it might be useful to work with transformed versions of the count data.
- The most obvious choice of transformation is the logarithm.
- (1) Variance stabilizing transformation (vst)
- (2) Regularized log transformation (rlog)
- This transforms the raw count data (which is heteroskedatic - variance grows with the mean) into homoskedatic data (variance is not dependant on the mean). 
- Both methods produce data on the log2 scale, and normalize for other factors such as library size. Setting `blind=TRUE` (the default) should be used to compare samples in a manner wholly unbiased about the information about experimental groups, for example to perform sample QC. 

- In order to test for differential expression, we operate on raw counts and use discrete distributions.


#### Variance stabilisting transformation (VST)
- VST by default uses a subset of 1000 rows to estimate the dispersion trend. This method is much faster than rlog, and is recommended if you have hundreds of samples.

```
vst <- vst(dds, blind = TRUE)
vst.data <- assay(vst)

# Regularized log (rlog) takes a long time with 50 or more samples
# rld <- rlog(dds, blind=FALSE)

```


#### Box plots 
- Box plots can be generated to check if one oe more samples are consistently higher than the others.
```
boxplot(vst.data,las=3, col="red")
```

#### Heatmap (Sample-to-sample distances)
- A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. 
- We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.
```
sampleDists <- dist(t(assay(vst)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vst$condition, vst$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

#### Differential expression using the DESeqDataSet object
- A DESeqDataSet object must have an associated design formula. The design formula expresses the variables which will be used in modeling. 
- The formula should be a tilde (~) followed by the variables with plus signs between them.
- The design can be changed later, however then all differential analysis steps should be repeated, as the design formula is used to estimate the dispersions and to estimate the log2 fold changes of the model.
- There are multiple ways of constructing a DESeqDataSet, depending on what pipeline was used upstream of DESeq2 to generated counts or estimated counts
- Here we have a matrix (as read in a dataframe above) of read counts prepared from our previous analysis using nfcore-rnaseq pipeline.
- So we use the function - DESeqDataSetFromMatrix

```
dds <- DESeqDataSetFromMatrix(countData = counttable,
                              colData = meta,
                              design = ~ condition)
# Check and discuss if required what is the warning!
#Warning in DESeqDataSet(se, design = design, ignoreRank) :
#  some variables in design formula are characters, converting to factors
```


### Pre-filtering of lowly expressed genes
- There are two reasons which make pre-filtering useful: 
- Genes with very low counts across all samples provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline.
- They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. 
- These genes should be filtered out prior to further analysis.

```
keep <- rowSums(cpm(counttable)>1) >=4
dds <- dds[keep,]
```

#### Explicitly set the factors levels 
- By default, R will choose a reference level for factors based on alphabetical order.
- So if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels.
- Setting the factor levels can be done in two ways
```
# Using factor
#dds$condition <- factor(dds$condition, levels = c("Wild","KO"))

#OR

# using relevel, just specifying the reference level:
dds$condition ~ relevel(dds$condition, ref="Wild")
```


### Principal component plot of the samples
- Related to the distance matrix is the PCA plot, which shows the samples in the 2D plane spanned by their first two principal components.
- This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects.

```
vsd <- vst(dds, blind=FALSE)
z<-plotPCA(vsd, intgroup=c("condition"))
z+ geom_text(aes_string(x = "PC1", y = "PC2", label = "name"),color = "black",size = 4)

# for the subsetted data
# There is an error with vsd (possibly due to very few rows/ genes with valid expression)
  # https://support.bioconductor.org/p/98634/
  # problem is that your expression matrix doesn't have more than 1000 genes (rows) in it.

# Using rlog (regularized log) transformation is slower but it works in this case for very few genes 
#rld <- rlog(dds)
#z<-plotPCA(rld, intgroup=c("condition"))
#z+ geom_text(aes_string(x = "PC1", y = "PC2", label = "name"), color = "black",size = 4)
```

### Scree plot
```
pca=prcomp(t(assay(vsd)),scale=FALSE)
options(repr.plot.width=0.5, repr.plot.height=0.5)
fviz_eig(pca, addlabels = TRUE)

# pca
#https://www.biostars.org/p/289196/
```

## Differential expression analysis

### The DESeq function
```

dds <- DESeq(dds)
res <- results(dds)
res


# padj 0.05
res_padj0.05<-results(dds,alpha=0.05)
summary(res_padj0.05)
resSig005_subset<-subset(res_padj0.05, padj < 0.05)
write.table(resSig005_subset, "res_DeSeq2_FDR0.05_comparison_Wild_vs_KO_FUllMatrix.tab", sep="\t", col.names=NA, quote=F)

# padj 0.1
res_padj0.1<-results(dds,alpha=0.1)
summary(res_padj0.1)
resSig01_subset<-subset(res_padj0.1, padj < 0.1)
write.table(resSig01_subset, "res_DeSeq2_FDR0.1_comparison_Wild_vs_KO_FUllMatrix.tab", sep="\t", col.names=NA, quote=F)

# Writing normalized counts
normalised_counts<-counts(dds,normalized=TRUE)
write.table(normalised_counts, "normalised_all_samples_DeSeq2_FUllMatrix.tab", sep="\t", col.names=NA, quote=F)

```

### Plot dispersion estimates
```
plotDispEsts(dds, ylim = c(1e-6,1e3) )
```

### Plot histogram of p-values
```
hist(resSig005_subset$pvalue, col = "lavender", main ="histogram of p-values" , xlab = "p-values")
```

### Tidy and annotate results
- Ordering by padj value
- Get gene names for ensembl IDs.

```
# https://github.com/stephenturner/annotables
# grch38 comes from library(annotables)
res_tidy.DE = tidy.DESeqResults(resSig005_subset)
res_tidy.DE <- res_tidy.DE %>% arrange(p.adjusted) %>% inner_join(grcm38, by = c(gene = "symbol")) %>% dplyr::select(gene,baseMean, estimate, stderror, statistic, p.value, p.adjusted) 
#res_tidy.DE
```

### Volcano plot
```
EnhancedVolcano(res_tidy.DE,
    lab = res_tidy.DE$gene,
    x = 'estimate',
    y = 'p.value',title = "Volvcano plot - WT versus KO",
    pointSize = 1.0,
    labSize = 5.0,
    pCutoff = 10e-16,
    FCcutoff = 1.5)

```



### GO over-representation analysis 
- Prepare input. 
- Filter for significant up and down regulated genes by P adjust and log fold change. 

```
# P adj < 0.05 
sig <- res_tidy.DE[res_tidy.DE$p.adjusted < 0.05, ]

# Upregulated: LFC > 1, remove NAs
sig.up <- sig[sig$estimate > 1, ]
sig.up <- na.omit(sig.up)
sig.up.LFC <- sig.up$estimate
names(sig.up.LFC) <- sig.up$gene
# Sort by LFC, decreasing
sig.up.LFC <- sort(sig.up.LFC, decreasing = TRUE)

# Downregulated: LFC < 1, remove NAs
sig.dn <- sig[sig$estimate < 1, ]
sig.dn <- na.omit(sig.dn)
sig.dn.LFC <- sig.dn$estimate
names(sig.dn.LFC) <- sig.dn$gene
# Sort by LFC, decreasing
sig.dn.LFC <- sort(sig.dn.LFC, decreasing = TRUE)
```



### Genes Down-regulated in WT


### GO over-representation analysis
- The clusterProfiler package implements enrichGO() for gene ontology over-representation test.
- Explanation 
- Observation 1
- Observation 2
```
ego.up <- enrichGO(gene = names(sig.up.LFC),
                      OrgDb = org.Mm.eg.db, 
                      keyType = 'SYMBOL',
                      readable = FALSE,
                      ont = "ALL",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)
```

### Bar plot
- Bar plot is the most widely used method to visualize enriched terms. It depicts the enrichment scores (e.g. p values) and gene count or ratio as bar height and color.
```
barplot(ego.up, showCategory=20)
```

### Dot plot
- A Dot plot is similar to a scatter plot and bar plot with the capability to encode another score as dot size.
- In R the dot plot displays the index (each category) in the vertical axis and the corresponding value in the horizontal axis, so you can see the value of each observation following a horizontal line from the label.
```{r, fig.height=7, fig.width=6}
dotplot(ego.up, showCategory=20,font.size = 10)
```

### cnetplot
- Both the barplot and dotplot only displayed most significant enriched terms, while users may want to know which genes are involved in these significant terms. The cnetplot depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network.
- The cnetplot depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network.
- Explanation 
- Observation 1
- Observation 2
```
cnetplot(ego.up, 
         categorySize="pvalue", 
         foldChange=sig.up.LFC,
         cex_label_gene = 1,
         showCategory = 5,cex_label_category=1.2,shadowtext='category')
```


### Heatmap-like functional classification
- The heatplot is similar to cnetplot, while displaying the relationships as a heatmap. 
- The gene-concept network may become too complicated if user want to show a large number significant terms. 
- The heatplot can simplify the result and more easy to identify expression patterns.
```
heatplot(ego.up)
```

### UpSet Plot
- The upsetplot is an alternative to cnetplot for visualizing the complex association between genes and gene sets. 
- It emphasizes the gene overlapping among different gene sets.
```
#Not working as of now!!! Check if this adds value?
#library("DOSE")
#upsetplot.library()
#upsetplot(ego.up,n=10)
```

## Genes Up-regulated in WT

### GO over-representation analysis
- Explanation 
- Observation 1
- Observation 2
```
ego.dn <- enrichGO(gene = names(sig.dn.LFC),
                      OrgDb = org.Mm.eg.db, 
                      keyType = 'SYMBOL',
                      readable = FALSE,
                      ont = "ALL",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.2)
```

### Bar plot
```
barplot(ego.dn, showCategory=20)
```

### Dot-plot
```
dotplot(ego.dn, showCategory=20,font.size = 10)
```


### cnetplot
- Explanation 
- Observation 1
- Observation 2
```
cnetplot(ego.dn, 
         categorySize="pvalue", 
         foldChange=sig.dn.LFC,
         cex_label_gene = 0.7,
         showCategory = 5,cex_label_category=1.5,shadowtext='category')
```


### Heatmap-like functional classification
```
heatplot(ego.dn)
```

### UpSet Plot
- The upsetplot is an alternative to cnetplot for visualizing the complex association between genes and gene sets. 
- It emphasizes the gene overlapping among different gene sets.
```
#Not working as of now!!! Check if this adds value?
#library("DOSE")
#upsetplot.library()
#upsetplot(ego.up,n=10)
```


