---
title: "Basic RNA-seq analysis workflow"
teaching: 10
exercises: 0
questions:
- "Have you seen a FASTQ file before?"
- "What does a FASTQ file contain?"
  

objectives:
- "Understand FASTQ format"
keypoints:
- XXX
---

#### RNA-Seq analysis workflow 
<figure>
<img src="{{ page.root }}/fig/rnaseq_workflow.png" style="margin:10px;height:350px"/>
  <figcaption> <b>Workflow : RNA-Seq analysis</b> </figcaption>
</figure><br> 

- Multiple steps are involved in the analysis of RNA-Seq data. 
- The first step is sequencing the reads. The format of the sequenced reads is called **[FASTQ](https://www.drive5.com/usearch/manual7/fastq_files.html)**.
- FASTQ files are text files containing sequence data with a quality (Phred) score for each base.
   
  <figure>
  <img src="{{ page.root }}/fig/fastq_format.jpg" style="margin:10px;height:350px"/>
  </figure><br>

- The above example shows a FASTQ read with 50nt in Illumina format (ASCII_BASE=33).
- There are always four lines per read.
- The four lines which represent a read in FASTQ are 
  - The first line starts with '@', followed by the label.
  - The second line contains the actual sequence.
  - The third line starts with '+'. In some variants, the '+' line contains a second copy of the label.
  - The fourth line contains the Q scores represented as ASCII characters.
- The quality score is an integer (Q) which is typically in the range 2 - 40


- Next step involves pre-processing for quality control (QC). A range of different tools can be used for performing an array of important function related tp data pre-processing for quality assessment etc.
- Next the good quality reads are aligning to a reference genome. 
- The number of reads mapped to each gene are then counted. This results in a table of counts, which is what we use to perform statistical analyses to determine differentially expressed genes and functionally important pathways.
- Each tool in the RNA-Seq pipeline is developed independantly and needs to be excecuted separately. 

