---
title: "RNA-Seq analysis: Two main steps"
teaching: 10
exercises: 0
questions:
- "How to convert sequencing data to gene counts?"
- "How to identify differentially expressed (DE) genes and enriched functions?"

objectives:
- "Two independant steps in RNA-Seq analysis workflow"
keypoints:
- XXX
---

This episode introduces the user to the two main excecution steps in RNA-Seq workflow as we approach it in this workshop

### RNA-Seq analysis workflow 
<figure>
  <img src="{{ page.root }}/fig/original_pipe_to_nextflowing.png" style="margin:10px;height:450px"/>
  <figcaption> RNA-Seq analysis workflow </figcaption>
</figure><br>

- RNA-seq is a popular method for simultaneously measuring and comparing the expression of tens of thousands of genes in multiple samples.
- There are many steps involved in analysing RNA-Seq data. 
- The analysis begins with sequencing reads (FASTQ files). These are usually aligned to a reference genome. The number of reads mapped to each gene can then be counted. This results in a table of counts, which is what we perform statistical analyses on to determine differentially expressed genes and pathways.
- 
