---
title: "RNA-Seq analysis: Two main steps"
teaching: 10
exercises: 0
questions:
- "How to convert sequencing data to gene counts?"
- "How to identify differentially expressed (DE) genes and enriched functions?"

objectives:
- "Two independant steps in RNA-Seq analysis workflow"
keypoints:
- XXX
---

This episode introduces the user to the two main excecution steps in RNA-Seq workflow as we approach it in this workshop

### RNA-Seq analysis workflow 
<figure>
  <img src="{{ page.root }}/fig/original_pipe_to_nextflowing.png" style="margin:10px"/>
  <figcaption> RNA-Seq analysis workflow </figcaption>
</figure><br>

- RNA-seq is a popular method for simultaneously measuring and comparing the expression of tens of thousands of genes in multiple samples.
- There are many steps involved in analysing RNA-Seq data, which can be categorised into two parts.
- **RNA-seq reads to counts**
  * The analysis begins with sequencing reads (FASTQ files).
  * Perform quality control of the reads using tools such as [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [Trim galore](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/).
  * The good quality reads are then usually aligned to a reference genome using tools such as [STAR](https://github.com/alexdobin/STAR) or [HiSAT2](http://daehwankimlab.github.io/hisat2/#:~:text=HISAT2%20is%20a%20fast%20and,to%20a%20single%20reference%20genome.). 
  * The number of reads mapped to each gene can then be counted. This results in a table of read-counts per gene.
- **RNA-seq counts to genes and functional enrichments**
  * The gene-counts can be used to perform statistical analyses to determine differentially expressed (DE) genes and pathways.
  * Multiple independant packages/libraries developed in R-programming can be used in this step.
  * [RStudio IDE](https://www.rstudio.com/) can be used to integrate the R code. 
  * R packages such as [DeSeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), [EdgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) etc are used for identification of differencially expressed (DE) genes.
  * R packages such as [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) and [pathfindR](https://cran.r-project.org/web/packages/pathfindR/index.html) and many others have been developed for functional enrichment analysis. 
  * In addition to R libraries, independant web-tools such as [PANTHER](http://www.pantherdb.org/), [DAVID](https://david.ncifcrf.gov/) and commercial tools such as [Ingenuity Pathway Analysis (IPA)](https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/?cmpid=QDI_GA_IPA&gclid=Cj0KCQjwkruVBhCHARIsACVIiOxh5HnHmJy72fRrmfQlnIhUpy2mpaFuVE2ClZDctskvzCoU-XCQerMaAmOhEALw_wcB) and [Metacore](https://clarivate.com/cortellis/webinars/metacore-integrated-pathway-analysis-for-multi-omics-data/) are also available.

