---
title: "Introduction to nfcore-rnaseq"
teaching: 10
exercises: 10
questions:
- "What is nfcore-rnaseq"

objectives:
- "Introduction to nextflow and nfcore"
keypoints:
- XXX
---

This episode gives an overview of the nfcore-rnaseq pipeline.

## nfcore-rnaseq

- nfcore-rnaseq is a bioinformatics pipeline that can be used to analyse RNA sequencing data obtained from organisms with a reference genome and annotation.
- The pipeline is built using [Nextflow](https://www.nextflow.io/)
-  It uses [Docker](https://www.docker.com/)/[Singularity](https://sylabs.io/) containers making installation trivial and results highly reproducible. 
-  Citation : doi: [10.5281/zenodo.1400710](https://doi.org/10.5281/zenodo.1400710)

# Workflow in nfcore-rnaseq

<figure>
  <img src="{{ page.root }}/fig/nf-core-rnaseq_metro_map_grey.png" style="margin:10px;height:300px"/>
  <figcaption> nfcore-rnaseq pipeline summary </figcaption>
</figure><br>

- nfcore-rnaseq uses [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/),[MultiQC](http://multiqc.info/), [Trim Galore!](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) and other tools for quality control and pre-processing of sequencing reads.
- The pipeline uses [STAR](https://github.com/alexdobin/STAR), [RSEM](https://github.com/deweylab/RSEM), [HISAT2](https://ccb.jhu.edu/software/hisat2/index.shtml) or [Salmon](https://combine-lab.github.io/salmon/) with gene/isoform counts for alignment and quantification.
- The pipeline uses [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) for initial exploratory analysis of the generated raw-counts matrix.


# What platform(s) to run the pipeline?
- Local machine?
- HPC (USyd-Artemis, NCI-GADI, Amazon cloud)
- VM (Pawsey Nimbus VM)

# What are the computations requirements for nfcore-rnaseq?

# What is the output from nfcore-rnaseq?

# Useful log outputs generated by nfcore-rnaseq

