---
title: "Pre-processing of read-countmatrix with R/RStudio"
teaching: 10
exercises: 10
questions:
- "How to import a count-matrix (output from nfcore) in R?"
- "How to generate PCA/Scree plots for exploring multivariate data?"
- "How to use the DESeq2 module to perform differential expression (DE) analysis?"
- "How to visualise the results from DeSeq2 analysis?" 


objectives:
- "Use RStudio on Nimbus instance to process the count-matrix and generate a list of differentially expressed genes."

keypoints:
- XXX
---

## Pre-processing steps

> ## Set the current working directory
> ```
> current_path <- getActiveDocumentContext()$path 
> setwd(dirname(current_path ))
> ```


> ## Load the R-libraries
> ```
> suppressMessages({
>   library("DESeq2")
>   library("edgeR")
>   library("limma")
>   library("RColorBrewer")
>   library("gplots")
>   library("ggplot2")
>   library("EnhancedVolcano")
>  library("factoextra")
>   library("devtools")
>   library("rstudioapi")
>   library("dplyr")
>   library("tibble")
>   library("tidyverse")
>   library("pheatmap")
>   library("biomaRt")
>   library("annotables")
>   library("org.Mm.eg.db")
>   library("biobroom")
>   library("clusterProfiler")
>   #library("pathfindR")
> })
> ```
> {: .language-bash}
{: .solution}


## Import the count matrix generated by nfcore-rnaseq
- The DESeq2 package expects count data from RNA-seq or another high-throughput sequencing experiment in the form of a matrix of integer values, as input. 
  - The value in the i-th row and the j-th column of the matrix tells how many reads can be assigned to gene i in sample j. 
  - The values in the matrix should be un-normalized counts or estimated counts of sequencing reads (for single-end RNA-seq) or fragments (for paired-end RNA-seq). 
  - The DESeq2 model internally corrects for library size, so transformed or normalized values such as counts scaled by library size should NOT be used as input.
- Next we read in the count matrix generated by the "nfcore-rnaseq" pipeline.
  - We have generated a matrix for a sub-set of the data from a specific region of the genome (chromosome 18). This was done with consideration to the time required for mapping and other processing steps involved.
  - We will be using this processed - subset matrix as part I of our analysis. This matrix should be available in the path "/".
  - As part II of the hands-on, we will redo the analysis with a FULL matrix which has been downloaded from the original manuscript. We can then use this full data for identifying gene functional categories of interest.
  - The file containing the full matrix should be available in the same folder "/" as "GSE81082_count_matrix_ENSIDs_symbols_nr.txt". 

> ## Import count matrix
> ```
> counttable_original<-read.delim("GSE81082_count_matrix_ENSIDs_symbols_nr.txt", header=T, row.names=1) 
> # View the count matrix
> #View(counttable_original)
> # Gene symbol as the identifier (when compared to ENSG ID)
> counttable<-counttable_original[,c("Symbol","WT1","WT2","WT3","KO1","KO2","KO3")]
> row.names(counttable) <- NULL
> # Convert Column  'GeneSymbol' to rowname)
> rownames(counttable) <- counttable$Symbol
> counttable<-counttable[,c("WT1","WT2","WT3","KO1","KO2","KO3")]
> #View(counttable)
> ```
> {: .language-bash}
{: .solution}



