---
title: "Run nfcore-rnaseq on a Nimbus instance"
teaching: 10
exercises: 0
questions:
- "Step 1: Convert the read-sequencing data into gene counts using nfcore-rnaseq pipeline"
- "Why use nfcore-rnaseq?"
- "Where to run nfcore-rnaseq pipeline?"
- "What are the commands to run nfcore-rnaseq?"

objectives:
- "Understanding how to run the nfcore-rnaseq on Nimbus"

keypoints:

---
This episode introduces the user to Step 1 in RNA-Seq analysis - Convert the read-sequencing data into gene counts using nfcore-rnaseq pipeline

### nfcore-rnaseq 
- Turn many steps into just a few.
- Flexibility in opting for tools at various steps of the analysis.
- Flexibility in opting for computational platforms and resources.


### Step1: Samplesheet (mandatory) input
You will need to create a [samplesheet](https://nf-co.re/rnaseq/3.7/usage#samplesheet-input) with information about the samples you would like to analyse before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 4 columns, and a header row as shown in the examples below. 

<figure>
  <img src="{{ page.root }}/fig/elaborate_samplesheet.png" style="margin:10px;height:150px"/>
</figure><br>

<figure>
  <img src="{{ page.root }}/fig/samplesheet_description.png" style="margin:10px;height:200px"/>
</figure><br>

The samplesheet required for today's analysis is a lot simpler, as we are using single-end reads. 
This samplesheet is already created and placed in the folder XXX.

> ## Identify strandedness
> Are the reads in our dataset single-end or paired-end?
> As far as we know this data is unstranded, but as a sanity check you can check the strandness. You can use RSeQC Infer Experiment tool to “guess” the strandness, as explained in the RNA-seq ref-based tutorial.
> ~~~
> 
> 
> ~~~
> <figure>
><img src="{{ page.root }}/fig/our_samplesheet.png" style="margin:10px;height:150px"/>
></figure><br>
> {: .language-bash}
{: .solution}



### Quick Start
~~~
nextflow run nf-core/rnaseq \
    --input samplesheet.csv \
    --outdir <OUTDIR> --genome GRCh38 
    -profile <docker/singularity/podman/shifter/charliecloud/conda/institute>
~~~

- Note that some form of configuration will be needed so that Nextflow knows how to fetch the required software. This is usually done in the form of a config profile. 
- The pipeline comes with config profiles such as docker, singularity, podman, shifter, charliecloud and conda which instruct the pipeline to use the named tool for software management.
- Please check [nf-core/configs](https://github.com/nf-core/configs#documentation) to see if a custom config file to run nf-core pipelines already exists for your Institute. If so, you can simply use -profile <institute> in your command. This will enable either docker or singularity and set the appropriate execution settings for your local compute environment.

### Pre-downloaded profiles
- We can select a specific profile such as **singularity**, and use the [nf-core](https://nf-co.re/tools/#downloading-pipelines-for-offline-use) download command to download images first, before running the pipeline. 
- Nextflow allows us to store and re-use the images from a central location for future pipeline runs.

### Actual nfcore-rnaseq command

~~~
base_path='DEFINE BASE PATH'

local_rnaseq_path='PATH_TO_DOWNLOADED_nfcore-rnaseq_IMAGE'

nextflow run $local_rnaseq_path/rnaseq-3.7 \
        --input samplesheet.csv \                                                      # samplesheet file-name
        --outdir results \                                                             # Define the name of the results folder
        --genome GRCm38 \                                                              # Reference genome
        -profile singularity \                                                         # profile e.g. singularity
        --fasta $base_path/Mouse_genomeFiles_from_sarek/genome.fa \                    # Path to the downloaded genome fasta file
        --gtf $base_path/Mouse_genomeFiles_from_sarek/sarek_Mmus.gtf \                 # Path to the downloaded 'gtf' file 
        --max_memory '64 GB' --max_cpus 2 \                                            # Define resources          
        --star_index $base_path/Mouse_genomeFiles_from_sarek/STARIndex_sarekGenome/ \  # Path to the downloaded genome 'STAR' mapper index file
        -with-report excecution_report_maxcpu2.html \                                  # Excecution log file-name (generated by nextflow)
        -with-timeline timeline_report_maxcpu2.html \                                  # Timeline log file-name.  (generated by nextflow)
        -r 3.4                        
~~~
  
  

