---
title: "Pre-processing of read-countmatrix with R/RStudio"
teaching: 10
exercises: 10
questions:
- "How to import a count-matrix (output from nfcore) in R?"
- "How to generate PCA/Scree plots for exploring multivariate data?"
- "How to use the DESeq2 module to perform differential expression (DE) analysis?"
- "How to visualise the results from DeSeq2 analysis?" 


objectives:
- "Use RStudio on Nimbus instance to process the count-matrix from nfcore-rnaseq and generate list of differentially expressed genes."
keypoints:
- XXX
---

This session shows us how to use the 'count-matrix' generated by the nfcore-rnaseq pipeline and run R programming in RStudio (Nimbus instance) to perform differential expression analysis. 


QC of count data
Multidimensional scaling plot
Density plots
Box plots
Voom variance plot
MD and Volcano plots for DE results


## Pre-processing steps

> ## Set the current working directory
> ~~~
> ## Set the current path as the current working directory
> current_path <- getActiveDocumentContext()$path 
> setwd(dirname(current_path ))
> print( getwd() )
> ~~~
> {: .language-bash}
{: .solution}

> ## Load the 'R' libraries
> ~~~
> ## Load all R packages (libraries) required for the various steps in this analysis.
> library("DESeq2")
> library("edgeR")
> library("limma")
> library("RColorBrewer")
> library("gplots")
> library("ggplot2")
> library("EnhancedVolcano")
> library("devtools")
> library("rstudioapi")
> library("dplyr")
> library("tibble")
> library("tidyverse")
> # Bioinformatics databases/formatting etc
> library("biomaRt")
> library("annotables")
> library("org.Mm.eg.db")
> library("biobroom")
> # Functional analysis
> library("clusterProfiler")
> library("pathfindR")
> ~~~
> {: .language-bash}
{: .solution}


> ## Read the count matrix 
> ~~~
> The countdata file contains information about genes (one gene per row)
> The first column has the gene symbol/gene ID and the remaining columns contain aligned read counts per sample
> # The combined-count matrix was generated using 'nfcore-rnaseq'
> counttable<-read.delim("GSE81082_count_matrix_ENSIDs_symbols_nr.txt", header=T, row.names=1)
> View(counttable)
> # ENSG as identifier
> subsetOfColumns <- c("WT1","WT2","WT3","KO1","KO2","KO3","Symbol")
> counttable <- counttable[subsetOfColumns]
> View(counttable)
> # Gene Symbol as identifier
> counttable<-counttable[,c("Symbol","WT1","WT2","WT3","KO1","KO2","KO3")]
> row.names(counttable) <- NULL
> #Column name (GeneSymbol to rowname)
> rownames(counttable) <- counttable$Symbol
> counttable<-counttable[,c("WT1","WT2","WT3","KO1","KO2","KO3")]
> View(counttable)
> ~~~
> {: .language-bash}
{: .solution}


### Filtering lowly expressed genes 
- It is recommended to filter for lowly expressed genes.
- The genes with very low coverage do not contribute to identification of differentially expressed genes.
- Rather, these geens can interfer with the statistical analysis.

> ~~~
> ## Filtering by minimum count-per-million 
> # Filter to remove lowly expressed genes 
> # Retain those with more than 1 counts per million in at least 4 (out of 6) samples 
> keep <-rowSums(cpm(counttable)>1) >=4
> data <-counttable[keep, ]
> dim(data)
> counttable<-data
> ~~~
> {: .language-bash}
{: .solution}

## Differential expression analysis using DESeq2.
> ~~~
> ## Make a design matrix
> meta <- data.frame(row.names=colnames(counttable),condition=c("Wild","Wild","Wild","KO","KO","KO"))
> meta$condition ~ relevel(meta$condition, ref="Wild")
> condition ~ relevel(factor(meta$condition), ref="Wild")
> dds<-DESeqDataSetFromMatrix(countData=counttable,colData=meta,design=~condition)
> dds <- estimateSizeFactors(dds)
> ~~~
> {: .language-bash}
{: .solution}

### Principal Component Analysis (PCA)
- Principal component analysis (PCA) is one among the many techniques adopted for exploring multivariate data like transcriptomes.
- PCA is an unsupervised analysis where we don’t need to specify the groups and it determines the greatest sources of variation in the data. 
- A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. 
- Ideally we expect that the greatest sources of variation in the data are the treatments/groups we are interested in. 
- It is also a useful tool for quality control and checking for outliers. 

> ## Transform data for PCA 
> ~~~
> vsd <- vst(dds, blind=FALSE)
> z<-plotPCA(vsd, intgroup=c("condition"))
> z+ geom_text(aes_string(x = "PC1", y = "PC2", label = "name"),color = "black",size = 4)
> # Writing normalised counts
> normalised_counts<-counts(dds,normalized=TRUE)
> write.table(normalised_counts, "normalised_counts_DeSeq2.tab", sep="\t", col.names=NA, quote=F)
> ~~~
> <figure>
>   <img src="{{ page.root }}/fig/PCA.png" style="margin:10px;height:350px"/>
> </figure><br>
> {: .language-bash}
{: .solution}

### Scree plot
- A scree plot is also produced that shows how much variation is attributed to each dimension. 
- e.g. In presence of a batch effect we could see high values for additional dimensions and will need to choose to include batch as an additional factor in the differential expression analysis. 

> ## Highlight the proportion of dimensions
> ~~~
> ~~~
> {: .language-bash}
{: .solution}

> ## Generate a list of DE genes
> ~~~
> # Compare Wild-type (WT) to Knockouts (KO)
> subsetOfColumns <- c("WT1","WT2","WT3","KO1","KO2","KO3")
> counttable_comparison_Wild_vs_KO_FULLMatrix <- counttable[subsetOfColumns]
> View(counttable_comparison_Wild_vs_KO_FULLMatrix)
> {: .language-bash}
{: .solution}


### DESeqDataSet object in DeSeq2
- The object class used by the DESeq2 package to store the read counts and the intermediate estimated quantities during statistical analysis is the DESeqDataSet, which will usually be represented in the code here as an object dds.
- A DESeqDataSet object must have an associated design formula. 
- The design formula expresses the variables which will be used in modeling. 
- The formula should be a tilde (~) followed by the variables with plus signs between them (it will be coerced into an formula if it is not already). 
- The design can be changed later, however then all differential analysis steps should be repeated, as the design formula is used to estimate the dispersions and to estimate the log2 fold changes of the model.

- There are different ways of constructing a DESeqDataSet, depending on what pipeline was used upstream of DESeq2 to generated counts or estimated counts ..
- DESeqDataSetFromMatrix can be used if you already have a matrix of read counts prepared from another source.
- Here we are using the count matrix generated using **nfcore-rnaseq.**
- To use DESeqDataSetFromMatrix, the user should provide the counts matrix, the information about the samples (the columns of the count matrix) as a DataFrame or data.frame, and the design formula.

> ## Identify differentially expressed genes with FDR cutoff 0.05
> ~~~
> meta_comparison_Wild_vs_KO_FULLMatrix <-data.frame(row.names=colnames(counttable_comparison_Wild_vs_KO_FULLMatrix),condition=c("Wild","Wild","Wild","KO","KO","KO"))
> meta_comparison_Wild_vs_KO_FULLMatrix$condition ~ relevel(meta_comparison_Wild_vs_KO_FULLMatrix$condition, ref="Wild")
> # Define condition
> condition ~ relevel(factor(meta_comparison_Wild_vs_KO_FULLMatrix$condition), ref="Wild")
> dds_comparison_Wild_vs_KO_FULLMatrix<-DESeqDataSetFromMatrix(countData=counttable_comparison_Wild_vs_KO_FULLMatrix,colData=meta_comparison_Wild_vs_KO_FULLMatrix,design=~condition)
> dds_comparison_Wild_vs_KO_FULLMatrix<-DESeq(dds_comparison_Wild_vs_KO_FULLMatrix)
> res_comparison_Wild_vs_KO_FULLMatrix<-results(dds_comparison_Wild_vs_KO_FULLMatrix,alpha=0.05)
> summary(res_comparison_Wild_vs_KO_FULLMatrix)
> res05_comparison_Wild_vs_KO_FULLMatrix<-sum(res_comparison_Wild_vs_KO_FULLMatrix$padj<0.05,na.rm=TRUE)
> resSig005_comparison_Wild_vs_KO_FULLMatrix<-subset(res_comparison_Wild_vs_KO_FULLMatrix, padj < 0.05)
> dim(resSig005_comparison_Wild_vs_KO_FULLMatrix)
> summary(resSig005_comparison_Wild_vs_KO_FULLMatrix)
> # Write results to a file
> write.table(resSig005_comparison_Wild_vs_KO_FULLMatrix, "res_DeSeq2_FDR0.05_comparison_Wild_vs_KO_FULL.tab", sep="\t", col.names=NA, quote=F)
> ~~~
> {: .language-bash}
{: .solution}

> ## Plot dispersion estimates
> ~~~
> plotDispEsts(dds_comparison_Wild_vs_KO_FULLMatrix, ylim = c(1e-6,1e3) )
> ~~~
> <figure>
>   <img src="{{ page.root }}/fig/Dispersion.png" style="margin:10px;height:350px"/>
> </figure><br>
> {: .language-bash}
{: .solution}

> ## Plot histogram of p-values
> ~~~
> #hist(res_comparison_Wild_vs_KO_FULLMatrix$pvalue, col = "lavender", main = title, xlab = "p-values")
> ~~~
> {: .language-bash}
{: .solution}

## Visualising results
- We might wish to look at the expression levels of certain genes of interest from the experiment perspective.
- We might also wish to visualise the expression of a few top genes from the differentially expressed genes (DE) gene list. 
- It will be good to observe if if the expressions of these genes are consistent amongst replicates with reference to the groups.

> ## Stripcharts of top genes
> ~~~

> ~~~
> {: .language-bash}
{: .solution}

## Heatmap of top genes
> ## Draw Heatmap
> ~~~

> ~~~

> {: .language-bash}
{: .solution}




