---
title: "nfcore-rnaseq countmatrix - pre-processing using RStudio"
teaching: 10
exercises: 10
questions:
- "How to read-in a count-matrix from nfcore-sarek in R?"
- "How to use the DESeq2 module to perform differential expression (DE) analysis"


objectives:
- "Use RStudio on Nimbus instance to process the count-matrix from nfcore-rnaseq and generate list of differentially expressed genes."
keypoints:
- XXX
---

This session shows us how to use the 'count-matrix' generated by the nfcore-rnaseq pipeline and run R programming in RStudio (Nimbus instance) to perform differential expression analysis. 


QC of count data
Multidimensional scaling plot
Density plots
Box plots
Voom variance plot
MD and Volcano plots for DE results


## Pre-processing steps

> ## Set the current working directory
> ~~~
> Set the current path as the current working directory
> current_path <- getActiveDocumentContext()$path 
> setwd(dirname(current_path ))
> print( getwd() )
> ~~~
> {: .language-bash}
{: .solution}

> ## Load the 'R' libraries
> ~~~
> library("DESeq2")
> library("edgeR")
> library("limma")
> library("RColorBrewer")
> library("gplots")
> library("ggplot2")
> library("EnhancedVolcano")
> library("devtools")
> library("rstudioapi")
> library("dplyr")
> library("tibble")
> library("tidyverse")
> # Bioinformatics databases/formatting etc
> library("biomaRt")
> library("annotables")
> library("org.Mm.eg.db")
> library("biobroom")
> # Functional analysis
> library("clusterProfiler")
> library("pathfindR")
> ~~~
> {: .language-bash}
{: .solution}


> ## Read the count matrix 
> ~~~
> # The combined-count matrix was generated using 'nfcore-rnaseq'
> counttable<-read.delim("GSE81082_count_matrix_ENSIDs_symbols_nr.txt", header=T, row.names=1)
> View(counttable)
> # ENSG as identifier
> subsetOfColumns <- c("WT1","WT2","WT3","KO1","KO2","KO3","Symbol")
> counttable <- counttable[subsetOfColumns]
> View(counttable)
> # Gene Symbol as identifier
> counttable<-counttable[,c("Symbol","WT1","WT2","WT3","KO1","KO2","KO3")]
> row.names(counttable) <- NULL
> #Column name (GeneSymbol to rowname)
> rownames(counttable) <- counttable$Symbol
> counttable<-counttable[,c("WT1","WT2","WT3","KO1","KO2","KO3")]
> View(counttable)
> ~~~
> {: .language-bash}
{: .solution}


> ## Filter to remove lowly expressed genes 
> ~~~
> ## Retian genes with more than 1 count-per-million in at least 4 (out of 6) samples  
> keep <-rowSums(cpm(counttable)>1) >=4
> data <-counttable[keep, ]
> dim(data)
> counttable<-data
> ~~~
> {: .language-bash}
{: .solution}

## Differential expression analysis using DESeq2.
> ~~~
> ## Make a design matrix
> meta <- data.frame(row.names=colnames(counttable),condition=c("Wild","Wild","Wild","KO","KO","KO"))
> meta$condition ~ relevel(meta$condition, ref="Wild")
> condition ~ relevel(factor(meta$condition), ref="Wild")
> dds<-DESeqDataSetFromMatrix(countData=counttable,colData=meta,design=~condition)
> dds <- estimateSizeFactors(dds)
> ~~~
> {: .language-bash}
{: .solution}

### Principal Component Analysis (PCA)
- Principal component analysis (PCA) is one among the many techniques adopted for exploring multivariate data like transcriptomes.
- PCA is an unsupervised analysis where we don’t need to specify the groups and it determines the greatest sources of variation in the data. 
- A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. 
- Ideally we expect that the greatest sources of variation in the data are the treatments/groups we are interested in. 
- It is also a useful tool for quality control and checking for outliers. 

> ## Transform data for PCA 
> ~~~
> vsd <- vst(dds, blind=FALSE)
> z<-plotPCA(vsd, intgroup=c("condition"))
> z+ geom_text(aes_string(x = "PC1", y = "PC2", label = "name"),color = "black",size = 4)
> # Writing normalised counts
> normalised_counts<-counts(dds,normalized=TRUE)
> write.table(normalised_counts, "normalised_counts_DeSeq2.tab", sep="\t", col.names=NA, quote=F)
> ~~~
> <figure>
>   <img src="{{ page.root }}/fig/PCA.png" style="margin:10px;height:350px"/>
> </figure><br>
> {: .language-bash}
{: .solution}

### Scree plot
- A scree plot is also produced that shows how much variation is attributed to each dimension. 
- e.g. In presence of a batch effect we could see high values for additional dimensions and will need to choose to include batch as an additional factor in the differential expression analysis. 

> ## Highlight the proportion of dimensions
> ~~~
> ~~~
> {: .language-bash}
{: .solution}

> ## Generate a list of DE genes
> ~~~
> # Compare Wild-type (WT) to Knockouts (KO)
> subsetOfColumns <- c("WT1","WT2","WT3","KO1","KO2","KO3")
> counttable_comparison_Wild_vs_KO_FULLMatrix <- counttable[subsetOfColumns]
> View(counttable_comparison_Wild_vs_KO_FULLMatrix)
> meta_comparison_Wild_vs_KO_FULLMatrix <-data.frame(row.names=colnames(counttable_comparison_Wild_vs_KO_FULLMatrix),condition=c("Wild","Wild","Wild","KO","KO","KO"))
> meta_comparison_Wild_vs_KO_FULLMatrix$condition ~ relevel(meta_comparison_Wild_vs_KO_FULLMatrix$condition, ref="Wild")
> # Define condition
> condition ~ relevel(factor(meta_comparison_Wild_vs_KO_FULLMatrix$condition), ref="Wild")
> dds_comparison_Wild_vs_KO_FULLMatrix<-DESeqDataSetFromMatrix(countData=counttable_comparison_Wild_vs_KO_FULLMatrix,colData=meta_comparison_Wild_vs_KO_FULLMatrix,design=~condition)
> dds_comparison_Wild_vs_KO_FULLMatrix<-DESeq(dds_comparison_Wild_vs_KO_FULLMatrix)
> res_comparison_Wild_vs_KO_FULLMatrix<-results(dds_comparison_Wild_vs_KO_FULLMatrix,alpha=0.05)
> summary(res_comparison_Wild_vs_KO_FULLMatrix)
> ~~~
> {: .language-bash}
{: .solution}

> ## Identify differentially expressed genes with FDR cutoff 0.05
> ~~~
> res05_comparison_Wild_vs_KO_FULLMatrix<-sum(res_comparison_Wild_vs_KO_FULLMatrix$padj<0.05,na.rm=TRUE)
> resSig005_comparison_Wild_vs_KO_FULLMatrix<-subset(res_comparison_Wild_vs_KO_FULLMatrix, padj < 0.05)
> dim(resSig005_comparison_Wild_vs_KO_FULLMatrix)
> summary(resSig005_comparison_Wild_vs_KO_FULLMatrix)
> # Write results to a file
> write.table(resSig005_comparison_Wild_vs_KO_FULLMatrix, "res_DeSeq2_FDR0.05_comparison_Wild_vs_KO_FULL.tab", sep="\t", col.names=NA, quote=F)
> ~~~
> {: .language-bash}
{: .solution}

> ## Plot dispersion estimates
> ~~~
> plotDispEsts(dds_comparison_Wild_vs_KO_FULLMatrix, ylim = c(1e-6,1e3) )
> ~~~
> <figure>
>   <img src="{{ page.root }}/fig/Dispersion.png" style="margin:10px;height:350px"/>
> </figure><br>
> {: .language-bash}
{: .solution}

> ## Plot histogram of p-values
> ~~~
> #hist(res_comparison_Wild_vs_KO_FULLMatrix$pvalue, col = "lavender", main = title, xlab = "p-values")
> ~~~
> {: .language-bash}
{: .solution}

## Visualising results
> ## Stripcharts of top genes
> {: .language-bash}
{: .solution}



